rasa/utils/tensorflow/rasa_layers.py

```diff
@@ -0,0 +1,757 @@
+        )
+
+        self._tf_layers["ffnn"] = layers.Ffnn(
+            config[HIDDEN_LAYERS_SIZES][attribute],
+            config[DROP_RATE],
+            config[REGULARIZATION_CONSTANT],
+            config[WEIGHT_SPARSITY],
+            layer_name_suffix=attribute,
+        )
+
+        # Prepare masking and computing helper variables for masked language modeling.
+        # This is only done for the text attribute and only if sequence-level (token-
+        # level) features are present (MLM requires token-level information).
+        if attribute == TEXT and SEQUENCE in attribute_signature and config[MASKED_LM]:
+            self.enables_mlm = True
+            self._tf_layers["mlm_input_mask"] = layers.InputMask()
+
+            # Unique IDs of different token types are needed to construct the possible
+            # label space for MLM. If dense features are present, they're used as such
+            # IDs, othwerise sparse features are embedded by a non-trainable
+            # DenseForSparse layer to create small embeddings that serve as IDs.
+            expect_dense_seq_features = any(
+                [not signature.is_sparse for signature in attribute_signature[SEQUENCE]]
+            )
+            if not expect_dense_seq_features:
+                self._tf_layers["sparse_to_dense_token_ids"] = layers.DenseForSparse(
+                    units=2,
+                    use_bias=False,
+                    trainable=False,
+                    name=f"sparse_to_dense_token_ids.{attribute}",
+                )
+        else:
+            self.enables_mlm = False
+
+        # Prepare the transformer
+        num_transformer_layers = config[NUM_TRANSFORMER_LAYERS]
+        if isinstance(num_transformer_layers, dict):
+            num_transformer_layers = num_transformer_layers[attribute]
+        transformer_size = config[TRANSFORMER_SIZE]
+        if isinstance(transformer_size, dict):
+            transformer_size = transformer_size[attribute]
+        if num_transformer_layers > 0:
+            self._tf_layers["transformer"] = TransformerEncoder(
+                num_layers=num_transformer_layers,
+                units=transformer_size,
+                num_heads=config[NUM_HEADS],
+                filter_units=transformer_size * 4,
+                reg_lambda=config[REGULARIZATION_CONSTANT],
+                dropout_rate=config[DROP_RATE],
+                attention_dropout_rate=config[DROP_RATE_ATTENTION],
+                sparsity=config[WEIGHT_SPARSITY],
+                unidirectional=config[UNIDIRECTIONAL_ENCODER],
+                use_key_relative_position=config[KEY_RELATIVE_ATTENTION],
+                use_value_relative_position=config[VALUE_RELATIVE_ATTENTION],
+                max_relative_position=config[MAX_RELATIVE_POSITION],
+                name=f"{attribute}_encoder",
+            )
+
+        # The output size of this layer depends on which component is the last one in
```

Comment:
let's create a separate helper to calculate `self.output_units`

Reply:
Yep.